caraml-store-serving/src/main/java/dev/caraml/serving/store/bigtable/HBaseOnlineRetriever.java

```diff
@@ -0,0 +1,173 @@
+                                  .position(rowCell.getFamilyOffset())
+                                  .limit(rowCell.getFamilyOffset() + rowCell.getFamilyLength())
+                                  .slice();
+                          String family = ByteString.copyFrom(familyBuffer).toStringUtf8();
+                          ByteString value = ByteString.copyFrom(valueBuffer);
+
+                          List<Feature> features;
+                          List<ServingServiceProto.FeatureReference> localFeatureReferences =
+                              featureReferences.stream()
+                                  .filter(
+                                      featureReference ->
+                                          featureReference.getFeatureTable().equals(family))
+                                  .collect(Collectors.toList());
+
+                          try {
+                            features =
+                                decodeFeatures(
+                                    tableName,
+                                    value,
+                                    localFeatureReferences,
+                                    reusedDecoder,
+                                    rowCell.getTimestamp());
+                          } catch (IOException e) {
+                            throw new RuntimeException("Failed to decode features from BigTable");
+                          }
+
+                          return features.stream();
+                        })
+                    .collect(Collectors.toList());
+              }
+            })
+        .collect(Collectors.toList());
+  }
+
+  @Override
+  public Map<ByteString, Result> getFeaturesFromSSTable(
+      String tableName, List<ByteString> rowKeys, List<String> columnFamilies) {
+    try {
+      Table table = this.client.getTable(TableName.valueOf(tableName));
+
+      // construct query get list
+      List<Get> queryGetList = new ArrayList<>();
+      rowKeys.forEach(
+          rowKey -> {
+            Get get = new Get(rowKey.toByteArray());
+            columnFamilies.forEach(cf -> get.addFamily(cf.getBytes()));
+
+            queryGetList.add(get);
+          });
+
+      // fetch data from table
+      Result[] rows = table.get(queryGetList);
+
+      // construct result
+      Map<ByteString, Result> result = new HashMap<>();
+      Arrays.stream(rows)
+          .filter(row -> !row.isEmpty())
+          .forEach(row -> result.put(ByteString.copyFrom(row.getRow()), row));
+
+      return result;
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  private List<Feature> decodeFeatures(
+      String tableName,
+      ByteString value,
+      List<ServingServiceProto.FeatureReference> featureReferences,
+      BinaryDecoder reusedDecoder,
+      long timestamp)
+      throws IOException {
+    ByteString schemaReferenceBytes = value.substring(0, 4);
```

Comment:
I understand that this is inspired from the current BigtableOnlineRetriever implementation, can we store the value of 4 in a variable with proper meaning, currently I don't understand why it uses 4 though

Reply:
Refactored out as a constant in BaseSchemaRegistry